core/ui/compose/common/src/main/kotlin/app/k9mail/core/ui/compose/common/mvi/UnidirectionalViewModel.kt

```diff
@@ -92,7 +91,7 @@ inline fun <reified STATE, EVENT, EFFECT> UnidirectionalViewModel<STATE, EVENT,
     val dispatch: (EVENT) -> Unit = { event(it) }
 
     LaunchedEffect(key1 = effect) {
-        effect.collectLatest {
+        effect.collect {
```

Comment:
There is no behavior change with regards to cancellation. The documentation of `collectLatest` succinctly states the difference between `collect` and `collectLatest`:

> The crucial difference from `collect` is that when the original flow emits a new value then the `action` block for the previous value is cancelled.

However, our `action` block only executes the non-suspending `handleEffect` block. So `collectLatest` cancelling the `action` block is effectively a no-op. The `handleEffect` block will always run to completion.

One guarantee we get from `SharedFlow.collect` (but not `Flow.collectLatest`) is this:
> It is guaranteed that, by the time the first suspension happens, `collect` has already subscribed to the `SharedFlow` and is eligible for receiving emissions.

This is necessary for `Effect.RequestCameraPermission` to be picked up when being triggered from `Event.StartScreen` (which is triggered during the first composition).

Side note: I don't see why we would ever want to have "intermediate" effects that could be skipped, especially when it comes to navigation. For things where intermediate states are not important we already have `State`.

Reply:
First of all architectural changes should be discussed prior and proven that changes the whole app behavior are tollerable. Instead of drive by changing it.

Reply:
I didn't consider this an architectural change. I always assumed effects are always executed to completion. That's how I understand the word "effect", that's how the code works, and that understanding seems to be consistent with how we use effects all throughout the app.

Reply:
I'd like to clarify the discussion around effects and side effects. From my perspective, a side effect in this context primarily relates to navigation and user interaction. I believe it's essential to consider the cost of navigation operations and how they impact the user experience. To minimize unnecessary actions, I think it's reasonable to assume that the last interaction is the desired one.

In the context of processing data, I understand that every emission can be considered a side effect. However, I'd like to focus on the navigation aspect and how we can optimize the user experience in this regard.

Reply:
After doing some research and testing, it appears that the name `collectLatest` is misleading. Unfortunately, the documentation doesn't provide much clarity on its capabilities, which might lead to misunderstandings.

The `collectLatest` operator in Kotlin's Flow API indeed cancels the previous block of work only if it contains suspendable operations. This is because non-suspendable work cannot be interrupted by the coroutine framework.

In the example from the docs, it's able to cancel the previous work:

```kotllin
flow {
    emit(1)
    delay(50)
    emit(2)
}.collectLatest { value ->
    println("Collecting $value")
   delay(100) // Emulate work
   println("$value collected")
}
```

Output: 

```
Collecting 1
Collecting 2
2 collected
```

However, if the block of work contains only non-suspendable operations, the collectLatest operator will not cancel the previous work. Instead, it will run the action block for every emission:

```kotlin
flow {
    emit(1)
    delay(50)
    emit(2)
}.collectLatest { value ->
    println("Collecting $value")
    println("$value collected")
}
```

Output:

```
Collecting 1
1 collected
Collecting 2
2 collected
```

In our case, the action block solely executes non-suspending code, so cancelling the action block using `collectLatest` is essentially a no-op. The handleEffect block will consistently run to completion.

Conclusion:

Before switching from `collectLatest` to `collect`, it's essential to check the affected code paths to ensure that all collecting blocks do not contain suspendable code blocks. Once confirmed, using `collect` is indeed better as it avoids the overhead.

Some documentation, articles or examples would have greatly helped to identify the issue more quickly and to avoid misunderstandings. For example `SharedFlow.collect` is different to `Flow.collect` but the IDE prefers importing the `Flow.collect` extension and opens it's documentation. But it's clear now.

For reference:
- [Kotlin Coroutine: collect](https://github.com/Kotlin/kotlinx.coroutines/blob/d8d6f8f37978b8e202d93b34f23f101df9c5724d/kotlinx-coroutines-core/common/src/flow/terminal/Collect.kt#L61-L97)
-[StackOverflow: An important detail missing](https://stackoverflow.com/a/78760108)